policies/engineering-principles.md

```diff
@@ -0,0 +1,50 @@
+
+
+### The Principle of Incremental Change
+
+To minimize the risk to the stability and security of the application, changes to the code should be incremental and as small as possible. Large changes to the codebase become difficult to review effectively and make the commit history less itemized. Furthermore, incremental change promotes better test coverage and makes it easier to require tests to cover the introduced changes.
+
+### The Principle of Incremental Decentralization
+
+We may sometimes ship with centralized services by default, but if we must, users should be able to opt out of any of these during onboarding, ideally easily swapping them out for decentralized alternatives. When we build features that rely upon centralized services we must build them in such a way where the experience isn't completely broken when the service is disrupted. We are trying to build a decentralized future and we must make choices that move us directionally towards those goals.
+
+### The Principle of Self-Soverign Identity
+
+We believe in the right of our users to own their own identity and data. We must go to great lengths to protect that right. Where we have features or services that might infringe upon that right we must treat those features as opt in, not opt out. When we do build features that make user data known to us we must not collect any personally identifiable information, and any data that can be tied to any chain must be collected in an anonymity protecting way. This principle must be extended to cover all third party code as well.
+
+### The Principle of Auditability
+
+All code contributions must be reviewed thoroughly for code quality, usability and adherence to these principles. Code contributions must be kept as small as possible (see Principle of Incremental Change) to promote auditability. Avoid long running feature branches because they are not auditable, and pose a risk to the stability and security of the product. We should engage in self-audit of code that we contribute, as well as actively engage in code review of others' pull requests to ensure best practices, and engineering principles are reflected in the work.
+
+
+### The Principle of "Be Methodical and Fix Things"
+
+We should prioritize fixing bugs and improving stability over creating new features whenever possible. All code generates new bugs and builds technical debt, this is inevitable. However, if we focus on building our features with stability in mind and throw out the "Move Fast and Break Things" mentality of most web 2.0 applications we can minimize the amount of bugs we introduce. Furthermore, whenever bugs are known to exist we should expend the necessary energy to remedy them. 
```

Comment:
Nit: This is fine as-is, but, I'm sure we could do better here. This doesn't present a convincing argument for _why_ we should "prioritize fixing bugs and improving stability", or paint a clear picture of what that entails.

On the "why", just saying that we can minimize bugs is underselling it. By being methodical and fixing things, we can maintain a more consistent developer velocity by minimizing disruptions caused by bugs, and we can improve velocity in the long-term by taking steps to avoid the buildup of technical debt. We can frame this principle as a commitment to _sustainable development_; to setting a pace that we can maintain years out, not just focusing on the current quarterly goals.

Then there's the user impact angle. In a sense, being methodical is being considerate of our users and the impact our work might have on them. And fixing things is a commitment to product quality.

On the "how", prioritizing fixing bugs and improving stability sounds great to me. Maybe we can say something more about the planning stage? On anticipating problems before they occur, having well thought out designs. Or maybe being methodical could refer to our efforts to maintain an auditable codebase. I'm less sure on this part. I guess we don't really need to spell out the "how" for any of these principles though, not here anyway.

Reply:
I agree that this doesn't go far enough in explaining the *why* (this is sort of a problem with all of these principles, but anyway). However, I agree with the conclusion of this paragraph. My vote is to expand this in a future PR.

Reply:
@mcmire We should be aiming to cover the why in these, at least at the surface level. I think fully covering the how is best left for supporting documents. That said, if there are principles that do not go far enough to cover the why please don't hesitate to point them out so we can get to a place that covers the why enough to get us into the supporting documents.

Reply:
Sounds good!

Reply:
I will take another stab at rewording this in a subsequent PR, we are at least agreed that it should be a principle we just need to do a better job detailing it.